File: Java/NapiJavaWrapper/src/com/nymi/api/wrapper/TransientNymiBandInfo.java
package com.nymi.api.wrapper;

import java.util.Vector;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

public class TransientNymiBandInfo {

	private JSONObject deviceInfo;
	
	public TransientNymiBandInfo(JSONObject jobj) {
		deviceInfo = jobj;
	}

	public Integer getRssiLast(){
	    try {
	        return deviceInfo.getInt("RSSI_last");
	    } catch (JSONException e) {
	    	return 0;
	    }
	}

	public Integer getRssiSmoothed(){
	    try {
	        return deviceInfo.getInt("RSSI_smoothed");
	    } catch (JSONException e) {
	    	return 0;
	    }
	}

	public Vector<String> getCommandsQueued() {
		Vector<String> commands = new Vector<String>();
		try {
	    	JSONArray cmdarr = deviceInfo.getJSONArray("commandQueue");
	        for (int i = 0 ; i < cmdarr.length() ; i++)
	            commands.add(cmdarr.getString(i));
		} catch (JSONException e) {
			return commands;
		}
        return commands;
	}

	public String getFirmwareVersion(){
	    try {
	        return deviceInfo.getString("firmwareVersion");
	    } catch (JSONException e) {
	    	return "";
	    }
	}

	public NymiJavaApi.FoundStatus getFoundState(){
	    try {
	        return NymiJavaApi.FoundStatus.valueOf(deviceInfo.getString("found").toUpperCase());
	    } catch (JSONException e) {
	    	return NymiJavaApi.FoundStatus.ERROR;
	    }
	}

	public NymiJavaApi.PresenceStatus getPresenceState(){
		try {
			return NymiJavaApi.PresenceStatus.valueOf("DEVICE_PRESENCE_" + deviceInfo.getString("present").toUpperCase());
		} catch(JSONException e) {
	    	return NymiJavaApi.PresenceStatus.ERROR;
		}
	}

	public Boolean isProvisioned(){		
	    if (deviceInfo.has("isProvisioned"))
	        return true;
	    else
	    	return false;
	}

	public double getSinceLastContact(){
	   try {
	        return deviceInfo.getDouble("sinceLastContact");
	   } catch (JSONException e) {
	    	return 0.0;
	   }
	}

	public double getAuthenticationWindowRemaining() {
		try {
			return deviceInfo.getJSONObject("provisioned").getDouble("authenticationWindowRemaining");
	    }
		catch (JSONException e) {
			return 0.0;
		}
	}

	public Integer getNumCommandsQueued(){
		try {
	    return deviceInfo.getJSONObject("provisioned").getInt("commandsQueued");
		} catch (JSONException e) {
			return 0;
		}
	}

	public Boolean enabledRoamingAuthentication(){
		try {
		    return deviceInfo.getJSONObject("provisioned").getBoolean("enabledRoamingAuthSetup");
			} catch (JSONException e) {
				return false;
			}
	}

	public Boolean enabledSigning(){
		try {
		    return deviceInfo.getJSONObject("provisioned").getBoolean("enabledSigning");
			} catch (JSONException e) {
				return false;
			}
	}

	public Boolean enabledSymmetricKeys(Boolean val){
		try {
		    return deviceInfo.getJSONObject("provisioned").getBoolean("enabledSymmetricKeys");
			} catch (JSONException e) {
				return false;
			}
	}
	
	public Boolean enabledTOTP(){
		try {
		    return deviceInfo.getJSONObject("provisioned").getBoolean("enabledTOTP");
			} catch (JSONException e) {
				return false;
			}
	}

	public String getPid(){
		try {
		    return deviceInfo.getJSONObject("provisioned").getString("pid");
			} catch (JSONException e) {
				return "";
			}
	}
	
}

File: Cplusplus/NapiCppWrapper/src/TransientNymiBandInfo.cpp
//
//  NymiDeviceInfo.cpp
//  NapiCpp
//
//  Created by Hanieh Bastani on 2016-04-27.
//  Copyright © 2016 Hanieh Bastani. All rights reserved.
//

#include "TransientNymiBandInfo.h"
#include "JsonUtilityFunctions.h"

TransientNymiBandInfo::TransientNymiBandInfo(nljson &jobj):deviceInfo(jobj){}

bool TransientNymiBandInfo::getRssiLast(int &val){
    
    nljson::iterator jit;
    if (hasKey(deviceInfo,{"RSSI_last"},jit)){
        val = jit.value();
        return true;
    }
    return false;
}

bool TransientNymiBandInfo::getRssiSmoothed(int &val){

    nljson::iterator jit;
    if (hasKey(deviceInfo,{"RSSI_smoothed"},jit)){
        val = jit.value();
        return true;
    }
    return false;
}

bool TransientNymiBandInfo::getCommandsQueued(std::vector<std::string> &commands){

    nljson::iterator jit;
    if (hasKey(deviceInfo,{"commandQueue"},jit)){
        for (auto cmd : jit.value()){
            commands.push_back(cmd);
        }
        return true;
    }
    return false;
}

bool TransientNymiBandInfo::getFirmwareVersion(std::string &val){

    nljson::iterator jit;
    if (hasKey(deviceInfo,{"firmwareVersion"},jit)){
        val = jit.value();
        return true;
    }
    val = "";
    return false;
}

bool TransientNymiBandInfo::getFoundState(FoundStatus &val){

    nljson::iterator jit;
    if (hasKey(deviceInfo,{"found"},jit)){
        std::string found = jit.value();
        if (foundEnum.find(found) != foundEnum.end()){
            val = foundEnum.find(found)->second;
            return true;
        }
    }
    val = FoundStatus::ERROR;
    return false;
}

bool TransientNymiBandInfo::getPresenceState(PresenceStatus &val){

    nljson::iterator jit;
    if (hasKey(deviceInfo,{"present"},jit)){
        std::string present = jit.value();
        if (presentEnum.find(present) != presentEnum.end()){
            val = presentEnum.find(present)->second;
            return true;
        }
    }
    val = PresenceStatus::ERROR;
    return false;
}

bool TransientNymiBandInfo::isProvisioned(bool &val){

    nljson::iterator jit;
    if (hasKey(deviceInfo,{"isProvisioned"},jit)){
        val = jit.value();
        return true;
    }
    return false;
}

bool TransientNymiBandInfo::getSinceLastContact(double &val){

    nljson::iterator jit;
    if (hasKey(deviceInfo,{"sinceLastContact"},jit)){
        val = jit.value();
        return true;
    }
    return false;
}

bool TransientNymiBandInfo::getAuthenticationWindowRemaining(double &val){

    nljson::iterator jit;
    if (hasKey(deviceInfo,{"provisioned","authenticationWindowRemaining"},jit)){
        val = jit.value();
        return true;
    }
    return false;
}

bool TransientNymiBandInfo::getNumCommandsQueued(int &val){

    nljson::iterator jit;
    if (hasKey(deviceInfo,{"provisioned","commandsQueued"},jit)){
        val = jit.value();
        return true;
    }
    return false;
}

bool TransientNymiBandInfo::enabledRoamingAuthentication(bool &val){
    
    nljson::iterator jit;
    if (hasKey(deviceInfo,{"provisioned","enabledRoamingAuthSetup"},jit)){
        val =  jit.value();
        return true;
    }
    return false;
}

bool TransientNymiBandInfo::enabledSigning(bool &val){

    nljson::iterator jit;
    if (hasKey(deviceInfo,{"provisioned","enabledSigning"},jit)){
        val = jit.value();
        return true;
    }
    return false;
}

bool TransientNymiBandInfo::enabledSymmetricKeys(bool &val){

    nljson::iterator jit;
    if (hasKey(deviceInfo,{"provisioned","enabledSymmetricKeys"},jit)){
        val = jit.value();
        return true;
    }
    return false;
}

bool TransientNymiBandInfo::enabledTOTP(bool &val){

    nljson::iterator jit;
    if (hasKey(deviceInfo,{"provisioned","enabledTOTP"},jit)){
        val = jit.value();
        return true;
    }
    return false;
}

bool TransientNymiBandInfo::getPid(std::string &val){

    nljson::iterator jit;
    if (hasKey(deviceInfo,{"provisioned","pid"},jit)){
        val = jit.value();
        return true;
    }
    val = "";
    return false;
}